### DISCUSSION TITLE
CMake 디버그 미세먼지 팁

### DISCUSSION THREAD

**Discussion by kuro11pow2** [euphony-io/euphony#55] [Knowledge Share]
![image](https://user-images.githubusercontent.com/47289893/134702699-a070761b-93f5-4323-b10c-dc3e56629082.png)

제가 사용중인 안드로이드 스튜디오 버전 + cmake 버전 3.10.2 에서는 ```CmakeLists.txt``` 에서 ```message()``` 함수를 사용해도 build output으로 출력이 제대로 나오지 않더라구요. 그래서 디버그에 어려움을 겪었는데요 ```message()``` 함수의 출력을 확인하는 방법 두 가지를 알아내어 공유드립니다.

## 로그 파일을 직접 열어서 확인
로그 파일 경로
```<euphony 루트>\euphony\.cxx\cmake\debug\arm64-v8a\cmake_server_log.txt```

![image](https://user-images.githubusercontent.com/47289893/134703140-b9ffe18f-a83e-4d88-98a2-87e3316e6b7f.png)
열어보면 위와 같이 상당히 어지러운 출력이 확인 되는데요 그래서 저는 알아보기 쉽게 아무 글자를 앞에 길게 넣어서 사용 중입니다.

```
execute_process(COMMAND ${ADB} shell getprop ro.product.cpu.abi
        OUTPUT_VARIABLE ADB_DEVICE_ABI
        RESULT_VARIABLE ADB_DEVICE_ABI_RESULT)

message(111111111111111111111 " " ADB_DEVICE_ABI  " " ${ADB_DEVICE_ABI})
message(111111111111111111111 " " ADB_DEVICE_ABI_RESULT " " ${ADB_DEVICE_ABI_RESULT})
```

## 터미널에서 명령어로 확인
다른 방법으로는 ```cmake_server_log.txt```가 존재하는 ```<euphony 루트>\euphony\.cxx\cmake\debug\arm64-v8a\``` 경로에서 ```cmake.exe```를 실행시키는 방법이 있습니다. ```cmake.exe```의 경로는 위 경로의 ```CMakeCache.txt```에 ```CMAKE_COMMAND``` 라는 변수로 저장되어있습니다. 

```${CMAKE_COMMAND}``` 로도 확인이 가능합니다. 

사용법을 정리하면 아래와 같습니다.
```cmd
>cd <euphony 루트>\euphony\.cxx\cmake\debug\arm64-v8a\
><CMAKE_COMMAND로 확인한 cmake.exe 경로> .
```
결과
![image](https://user-images.githubusercontent.com/47289893/134705001-f6b94812-c137-4ea1-af58-34a4b50eb1ec.png)
![image](https://user-images.githubusercontent.com/47289893/134705291-195c856e-6328-4fcb-844b-9830c96a6b4f.png)
![image](https://user-images.githubusercontent.com/47289893/134706483-ce33c5ff-45ab-4b4f-b7cb-69f9a2f20266.png)

명령어를 사용하는 방법은 경로에 들어가기가 다소 번거롭긴 하지만 ```message()```로 출력한 부분은 물론 ```execute_process()```로 실행한 명령의 출력도 보여줘서 사용하기 좋더라구요.

아쉬운 점은 ```euphony\src\main\cpp\tests\CMakeLists.txt``` 파일의 마지막 부분을 보면 cmake.run.test.script 를 실행하는데
```
# Run gtest & get the result.
add_custom_command(TARGET ${TEST_EUPHONY} POST_BUILD
        COMMAND ${CMAKE_COMMAND} -DADB:STRING="${ADB}" -DTARGET_TEST_LIB_DIR:STRING="${TARGET_TEST_LIB_DIR}" -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake.run.test.script
                )
```

```
execute_process(COMMAND ${ADB} shell LD_LIBRARY_PATH=${TARGET_TEST_LIB_DIR} ${TARGET_TEST_LIB_DIR}/testEuphony
                OUTPUT_VARIABLE GTEST_EUPHONY_OUTPUT
                RESULT_VARIABLE GTEST_RESULT
                ERROR_VARIABLE GTEST_ERROR_OUTPUT
                )

if( ${GTEST_RESULT} GREATER 0 )
    string(STRIP ${GTEST_EUPHONY_OUTPUT} GTEST_EUPHONY_OUTPUT)
    string(REPLACE "\n\n" "\r[NEXT]\r" GTEST_EUPHONY_OUTPUT ${GTEST_EUPHONY_OUTPUT})
    string(REPLACE "\n" "\r" GTEST_EUPHONY_OUTPUT ${GTEST_EUPHONY_OUTPUT})
    message(FATAL_ERROR "** Gtest Failure (${GTEST_RESULT}) **\r${GTEST_EUPHONY_OUTPUT}")
else()
    message("**** Gtest Success ****")
endif()
````
여기서 호출한 message의 출력은 안 나오네요 ㅠㅠ

**Comment by seohy02** [comment]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ미세먼지 팁이라니,,,, 완전 꿀팁입니다  감사해요!👍👍👍👍

**Reply by kuro11pow2** [reply]
감사합니다 ㅎㅎ

**Comment by ycs1m1yk** [comment]
아주 도움됐습니다!!

cmake.run.test.script의 message들은 여기 나오긴합니다ㅎㅎ
`${CMAKE_CACHEFILE_DIR}\\android_gradle_build_stdout_targets.txt`

### CASE 1:  Gtest Success

![build_stdout_case_success](https://user-images.githubusercontent.com/27720475/134766106-e6da9c6f-fcd5-4164-9aad-28b98347ea3e.png)


### CASE 2:  Gtest Failure

![build_stdout_case_failure](https://user-images.githubusercontent.com/27720475/134766129-34b76c24-ea02-449e-a715-5ad2c8b3c16f.png)

**Reply by kuro11pow2** [reply]
좋은 정보 감사합니다!! 혹시 저도 해당 파일에 출력이 남는지 확인해봤는데 아쉽게도 텅 비어있네요 ㅠㅠ
혹시 flag라든지 gradle 설정을 수정해야 나오는 걸까요??

![image](https://user-images.githubusercontent.com/47289893/134766427-0872e04b-d8d5-41a5-984e-589b4c064f87.png)

**Reply by yeonns** [reply]
@kuro11pow2 기기연결 후 빌드가 최초한번 성공하면 저 파일에 다 기록되는데, 한번 더 빌드를 실행하면 이전에 성공한 빌드기록이 남아서 gtest를 수행하지 않기때문에 파일이 비어있는 걸 저도 확인했어요! 아마 .cxx 디렉토리를 전부삭제 후, 다시빌드하면 결과볼수있을 거 같습니다!

**Reply by kuro11pow2** [reply]
@yeonns .cxx 디렉토리를 지우고 다시 빌드해보니까 결과를 확인할 수 있네요! 알려주셔서 감사합니다!

**Reply by ycs1m1yk** [reply]
https://github.com/euphony-io/euphony/issues/46#issuecomment-926623407
46번 이슈에 영석님 코멘트대로

1. `build.gradle(:euphony)`에 BUILD_CACHE 라는 cmake 인자를 설정하고,

2. `...\\cpp\\tests\\CMakeList.txt` 에 아래와 같은 코드 한 줄 추가한 상태입니다!
> ```cmake
> file(REMOVE ${BUILD_CACHE})
> ```
